### ISSUE TITLE
assert in ProcessWaitState on Linux arm64 

### ISSUE THREAD

**Issue by wfurt** [dotnet/runtime#74795]
Disabled tests:
- 9/8 - Mono Linux arm64 in PR #75266 - see https://github.com/dotnet/runtime/issues/74795#issuecomment-1241036585

Crash in **System.Net.Requests.Tests** Work Item - Last [7 days](https://runfo.azurewebsites.net/search/tests/?q=started%3A%7E7+definition%3Aruntime+name%3A%22system.net.requests.tests+work+item%22) in Runfo:
- 9/8 - 142x failures out of which 115x are on arm64-Mono

Day | Run | Details
-- | -- | --
9/7 | Rolling run [6968](https://runfo.azurewebsites.net/view/build/?number=6968) | Core Dump - net7.0-Linux-Release-**arm64-NativeAOT**_Release-(Ubuntu.1804.ArmArch.Open)Ubuntu.1804.ArmArch.Open
9/6 | Rolling run [6173](https://runfo.azurewebsites.net/view/build/?number=6173) | 2x Core Dump - net7.0-Linux-Release-**arm64-Mono**_release-(Ubuntu.1804.ArmArch.Open)Ubuntu.1804.ArmArch.Open & net7.0-Linux-Release-**arm64-Mono**_release-(Debian.11.Arm64.Open)Ubuntu.1804.Armarch.Open
9/4 | Rolling run [4317](https://runfo.azurewebsites.net/view/build/?number=4317) | Code Rump - net7.0-Linux-Release-**arm64-CoreCLR**_release-(Ubuntu.2204.Arm64.Open)Ubuntu.1804.ArmArch.Open
8/30 | Rolling run [1975580](https://runfo.azurewebsites.net/view/build/?number=1975580) | 2x Core Dump - net7.0-Linux-Release-**arm64-Mono**_release-(Ubuntu.1804.ArmArch.Open)Ubuntu.1804.ArmArch.Open & net7.0-Linux-Release-**arm64-Mono**_release-(Debian.11.Arm64.Open)Ubuntu.1804.Armarch.Open
8/29 | Rolling run [1973318](https://runfo.azurewebsites.net/view/build/?number=1973318) | Code Dump - net7.0-Linux-Release-**arm64-Mono**_release-(Ubuntu.1804.ArmArch.Open)Ubuntu.1804.ArmArch.Open
8/28 | Rolling run [1972808](https://runfo.azurewebsites.net/view/build/?number=1972808) | [Core Dump](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-main-acee0b7c07dd4330b5/System.Net.Requests.Tests/1/core.1001.21?%3Fhelixlogtype%3Dresult) - net7.0-Linux-Release-**arm64-Mono**_release-(Debian.11.Arm64.Open)Ubuntu.1804.Armarch.Open
8/28 | Rolling run [1972465](https://runfo.azurewebsites.net/view/build/?number=1972465) | Core Dump - net7.0-Linux-Release-**arm64-Mono**_release-(Ubuntu.1804.ArmArch.Open)Ubuntu.1804.ArmArch.Open
8/26 | Rolling run [1970381](https://runfo.azurewebsites.net/view/build/?number=1970381) | [Core Dump](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-main-aaacaf8e0a7f46c4ad/System.Net.Requests.Tests/1/core.1001.21?%3Fhelixlogtype%3Dresult) - net7.0-Linux-Release-**arm64-CoreCLR**_release-(Ubuntu.2204.Arm64.Open)Ubuntu.1804.ArmArch.Open

Related to #69125.
https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-main-aaacaf8e0a7f46c4ad/System.Net.Requests.Tests/1/console.1429bd54.log?%3Fhelixlogtype%3Dresult

```
/root/helix/work/workitem/e /root/helix/work/workitem/e
  Discovering: System.Net.Requests.Tests (method display = ClassAndMethod, method display options = None)
  Discovered:  System.Net.Requests.Tests (found 349 of 367 test cases)
  Starting:    System.Net.Requests.Tests (parallel test collections = on, max threads = 4)
Process terminated. Error while reaping child. errno = 10
   at System.Environment.FailFast(System.String)
   at System.Diagnostics.ProcessWaitState.TryReapChild(Boolean)
   at System.Diagnostics.ProcessWaitState.CheckChildren(Boolean, Boolean)
   at System.Diagnostics.Process.OnSigChild(Int32, Int32)
./RunTests.sh: line 168:    21 Aborted                 (core dumped) "$RUNTIME_PATH/dotnet" exec --runtimeconfig System.Net.Requests.Tests.runtimeconfig.json --depsfile System.Net.Requests.Tests.deps.json xunit.console.dll System.Net.Requests.Tests.dll -xml testResults.xml -nologo -nocolor -notrait category=IgnoreForCI -notrait category=OuterLoop -notrait category=failing $RSP_FILE
```
dump: https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-main-aaacaf8e0a7f46c4ad/System.Net.Requests.Tests/1/core.1001.21

errno 10 -> `ENOCHILD`

```
> clrstack -a
OS Thread Id: 0x1c (0)
        Child SP               IP Call Site
0000007EBADC5460 0000007fa357f200 [HelperMethodFrame_1OBJ: 0000007ebadc5460] System.Environment.FailFast(System.String)
0000007EBADC55E0 0000007F64861C34 System.Diagnostics.ProcessWaitState.TryReapChild(Boolean) [/_/src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessWaitState.Unix.cs @ 589]
    PARAMETERS:
        this (0x0000007EBADC5630) = 0x000000156f599f80
        configureConsole (0x0000007EBADC562C) = 0x0000000000000001
    LOCALS:
        0x0000007EBADC5620 = 0x000000156f599fc8
        0x0000007EBADC5618 = 0x0000000000000001
        0x0000007EBADC5610 = 0x0000000060230000
        0x0000007EBADC560C = 0x00000000ffffffff
        0x0000007EBADC5608 = 0x0000000000000001
        0x0000007EBADC5600 = 0x000000000000000a

0000007EBADC5640 0000007F648610C8 System.Diagnostics.ProcessWaitState.CheckChildren(Boolean, Boolean) [/_/src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessWaitState.Unix.cs @ 614]
    PARAMETERS:
        reapAll (0x0000007EBADC5724) = 0x0000000000000000
        configureConsole (0x0000007EBADC5720) = 0x0000000000000001
    LOCALS:
        0x0000007EBADC5718 = 0x000000156f42a280
        0x0000007EBADC5710 = 0x0000000000000001
        0x0000007EBADC570C = 0x0000000000000000
        0x0000007EBADC5708 = 0x000000000000003c
        0x0000007EBADC5700 = 0x000000156f599f80
        0x0000007EBADC56F8 = 0x0000000000000000
        0x0000007EBADC56F0 = 0x0000000000000000
        0x0000007EBADC56E8 = 0x0000000000000000
        0x0000007EBADC56C0 = 0x0000000000000000
        0x0000007EBADC56B0 = 0x0000000000000000
        0x0000007EBADC56A8 = 0x0000000000000000
        0x0000007EBADC5690 = 0x0000000000000000
        0x0000007EBADC5688 = 0x0000000000000000
        0x0000007EBADC5680 = 0x0000000000000000

0000007EBADC5730 0000007F64860B40 System.Diagnostics.Process.OnSigChild(Int32, Int32) [/_/src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs @ 1104]
    PARAMETERS:
        reapAll = <no data>
        configureConsole (<CLR reg>) = 0x0000000000000001
    LOCALS:
        <CLR reg> = 0x0000000000000001
        <no data>
        <no data>

> dumpobj 0x000000156f599f80
Name:        System.Diagnostics.ProcessWaitState
MethodTable: 0000007f6480b9a8
EEClass:     0000007f64827760
Tracked Type: false
Size:        72(0x48) bytes
File:        /root/helix/work/correlation/shared/Microsoft.NETCore.App/8.0.0/System.Diagnostics.Process.dll
Fields:
              MT    Field   Offset                 Type VT     Attr            Value Name
0000007f62d49320  4000179        8        System.Object  0 instance 000000156f599fc8 _gate
0000007f63b4e388  400017a       20         System.Int32  1 instance               60 _processId
0000007f63b4ae88  400017b       28       System.Boolean  1 instance                1 _isChild
0000007f63b4ae88  400017c       29       System.Boolean  1 instance                1 _usesTerminal
0000007f63e07350  400017d       10 ...eading.Tasks.Task  0 instance 0000000000000000 _waitInProgress
0000007f63b4e388  400017e       24         System.Int32  1 instance                2 _outstandingRefCount
0000007f63b4ae88  400017f       2a       System.Boolean  1 instance                0 _exited
0000007f63c7cc10  4000180       2c ...Private.CoreLib]]  1 instance 000000156f599fac _exitCode
0000007f63f72c08  4000181       38      System.DateTime  1 instance 000000156f599fb8 _exitTime
0000007f645345b0  4000182       18 ....ManualResetEvent  0 instance 000000156f599fe0 _exitedEvent
0000007f6480d118  4000177       60 ...nostics.Process]]  0   static 000000156f42a230 s_processWaitStates
0000007f6480d118  4000178       68 ...nostics.Process]]  0   static 000000156f42a280 s_childProcessWaitStates
```

and with (some) symbols
```
> clrstack -i -a



Dumping managed stack and managed variables using ICorDebug.
=============================================================================
Child SP         IP               Call Site
0000007EBADC4A60 0000007fa357f200 [NativeStackFrame]
0000007EBADC5460 (null) [Internal call: 0000007EBADC5460]
0000007EBADC55E0 0000007f64861c34 [DEFAULT] [hasThis] Boolean System.Diagnostics.ProcessWaitState.TryReapChild(Boolean) (/root/helix/work/correlation/shared/Microsoft.NETCore.App/8.0.0/System.Diagnostics.Process.dll)

PARAMETERS:
  + System.Diagnostics.ProcessWaitState this @ 0x156f599f80
  + bool configureConsole  = true

LOCALS:
  + (Error 0x80004005 retrieving local variable 'local_0')
  + (Error 0x80004005 retrieving local variable 'local_1')
  + int exitCode  = 1612906496
  + int waitResult  = -1
  + (Error 0x80004005 retrieving local variable 'local_4')
  + int errorCode  = 10

0000007EBADC5640 0000007f648610c8 [DEFAULT] Void System.Diagnostics.ProcessWaitState.CheckChildren(Boolean,Boolean) (/root/helix/work/correlation/shared/Microsoft.NETCore.App/8.0.0/System.Diagnostics.Process.dll)

PARAMETERS:
  + bool reapAll  = false
  + bool configureConsole  = true

LOCALS:
  + (Error 0x80004005 retrieving local variable 'local_0')
  + (Error 0x80004005 retrieving local variable 'local_1')
  + bool checkAll  = false
  + int pid  = 60
  + System.Diagnostics.ProcessWaitState pws @ 0x156f599f80
  + int errorCode  = 0
  + System.Diagnostics.ProcessWaitState firstToRemove = null
  + System.Collections.Generic.List`1&lt;System.Diagnostics.ProcessWaitState&gt; additionalToRemove = null
  + (Error 0x80004005 retrieving local variable 'local_8')
  + System.Collections.Generic.KeyValuePair`2&lt;int,System.Diagnostics.ProcessWaitState&gt; kv @ 0x7ebadc56b0
  + System.Diagnostics.ProcessWaitState pws = null
  + (Error 0x80004005 retrieving local variable 'local_11')
  + System.Diagnostics.ProcessWaitState pws = null
  + (Error 0x80004005 retrieving local variable 'local_13')

0000007EBADC5730 0000007f64860b40 [DEFAULT] I4 System.Diagnostics.Process.OnSigChild(I4,I4) (/root/helix/work/correlation/shared/Microsoft.NETCore.App/8.0.0/System.Diagnostics.Process.dll)

PARAMETERS:
  + (Error 0x80131304 retrieving parameter 'reapAll')
  + int configureConsole  = 1

LOCALS:
  + bool childrenUsingTerminalPre  = true
  + (Error 0x80004005 retrieving local variable 'childrenUsingTerminalPost')
  + (Error 0x80004005 retrieving local variable 'local_2')

0000007EBADC5790 0000007f60218260 [NativeStackFrame]
Stack walk complete.
=============================================================================
```

cc: @tmds

```json 
{ "ErrorMessage":"Error while reaping child. errno = 10" } 
```

#### Summary
|24-Hour Hit Count|7-Day Hit Count|1-Month Count|
|---|---|---|
|0|0|0|

**Comment by tmds** [comment]
We had this last year for `System.Net.Requests.Tests` too: https://github.com/dotnet/runtime/issues/55618.

This `FailFast` happens when some other code in the process reaps a child process that was started by .NET.

**Comment by danmoseley** [comment]
In such a case, it's a fail fast necessary? Would it be reasonable to just continue?

**Comment by tmds** [comment]
Someone proposed such a change. We didn't take it because we obtain the exit code when we reap the child (from `waitpid`). So this affects the further state of the program. (https://github.com/dotnet/runtime/pull/70705#issuecomment-1156717739).

**Comment by danmoseley** [comment]
> Someone proposed such a change. We didn't take it because we obtain the exit code when we reap the child (from waitpid). So this affects the further state of the program. (https://github.com/dotnet/runtime/pull/70705#issuecomment-1156717739).

I read that comment as a general statement that we prefer to fail fast than continue in a corrupted or undefined state (cc @jkotas). In this case, can we reasonably recover from this error? It seems we use the error code to release resources -- can we just do that? Or is a leak inevitable here?

It seems fragile to terminate the app suddenly in the case some child process terminates or is killed for some reason. I guess this does not happen on Windows.

**Comment by wfurt** [comment]
One another questions is what this tart happening now. It can certainly be race condition that always existed. I tried to reproduce it locally (on ARM64) but so far no luck.  
Only change I'm aware of is we updated test images to include MsQuic. That really should not impact this test set but timing may changed. I did quick search through MsQuic code and I did not see any attempts to deal with process management. 
So the question is is what is reaping the process...

**Comment by jkotas** [comment]
> than continue in a corrupted or undefined state (cc @jkotas).

The corrupted state here is akin to calling `free` on a pointer that you do not own.

The situation is not triggered by child process terminating or being killed for some reason. The situation is triggered by some code in the current process handling it incorrectly.

It is also possible that the root cause has nothing to do with process reaping. Instead, the root cause may be a memory corruption somewhere else that this crash is a side-effect of.

**Comment by wfurt** [comment]
I did check the test suites and I don't see any other attempts to reap or execute besides our `RemoteExecutor` @tmds. 
```
16623 execve("/media/5bfe43cb-dd6e-413b-8b94-00f69f90ecc2/wfurt/runtime/artifacts/bin/testhost/net7.0-Linux-Debug-arm64/dotnet", ["/media/5bfe43cb-dd6e-413b-8b94-0"..., "exec", "--runtimeconfig", "System.Net.Requests.Tests.runtim"..., "--depsfile", "System.Net.Requests.Tests.deps.j"..., "xunit.console.dll", "System.Net.Requests.Tests.dll", "-xml", "testResults.xml", "-nologo", "-notrait", "category=OuterLoop", "-notrait", "category=failing", "-nocolor"], 0x7fe0751af8 /* 13 vars */) = 0
16644 execve("/mnt/runtime/artifacts/bin/testhost/net7.0-Linux-Debug-arm64/dotnet", ["/mnt/runtime/artifacts/bin/testh"..., "exec", "--runtimeconfig", "/mnt/runtime/artifacts/bin/Syste"..., "--depsfile", "/mnt/runtime/artifacts/bin/Syste"..., "/mnt/runtime/artifacts/bin/Syste"..., "System.Net.Requests.Tests, Versi"..., "System.Net.Tests.WebRequestTest+"..., "<DefaultWebProxy_SetThenGet_Valu"..., "/tmp/31jqjssf.ryi"], 0x7e8803b9a0 /* 13 vars */) = 0
16651 execve("/mnt/runtime/artifacts/bin/testhost/net7.0-Linux-Debug-arm64/dotnet", ["/mnt/runtime/artifacts/bin/testh"..., "exec", "--runtimeconfig", "/mnt/runtime/artifacts/bin/Syste"..., "--depsfile", "/mnt/runtime/artifacts/bin/Syste"..., "/mnt/runtime/artifacts/bin/Syste"..., "System.Net.Requests.Tests, Versi"..., "System.Net.Tests.GlobalProxySele"..., "<Select_Success>b__1_0", "/tmp/0rb3jbxa.g4r"], 0x7e90005e40 /* 13 vars */) = 0
16640 --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=16644, si_uid=1001, si_status=42, si_utime=10, si_stime=6} ---
16630 waitid(P_ALL, 0, {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=16644, si_uid=1001, si_status=42, si_utime=0, si_stime=0}, WNOHANG|WEXITED|WNOWAIT, NULL) = 0
16630 wait4(16644, [{WIFEXITED(s) && WEXITSTATUS(s) == 42}], WNOHANG, NULL) = 16644
16630 waitid(P_ALL, 0, {}, WNOHANG|WEXITED|WNOWAIT, NULL) = 0
16638 --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=16651, si_uid=1001, si_status=42, si_utime=18, si_stime=10} ---
16663 execve("/mnt/runtime/artifacts/bin/testhost/net7.0-Linux-Debug-arm64/dotnet", ["/mnt/runtime/artifacts/bin/testh"..., "exec", "--runtimeconfig", "/mnt/runtime/artifacts/bin/Syste"..., "--depsfile", "/mnt/runtime/artifacts/bin/Syste"..., "/mnt/runtime/artifacts/bin/Syste"..., "System.Net.Requests.Tests, Versi"..., "System.Net.Tests.WebRequestTest+"..., "<SendGetRequest_WithGlobalCacheP"..., "/tmp/dkgtmmfz.ado"], 0x7e8803d100 /* 13 vars */) = 0
16630 waitid(P_ALL, 0, {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=16651, si_uid=1001, si_status=42, si_utime=0, si_stime=0}, WNOHANG|WEXITED|WNOWAIT, NULL) = 0
16630 wait4(16651, [{WIFEXITED(s) && WEXITSTATUS(s) == 42}], WNOHANG, NULL) = 16651
16630 waitid(P_ALL, 0, {}, WNOHANG|WEXITED|WNOWAIT, NULL) = 0
16664 execve("/mnt/runtime/artifacts/bin/testhost/net7.0-Linux-Debug-arm64/dotnet", ["/mnt/runtime/artifacts/bin/testh"..., "exec", "--runtimeconfig", "/mnt/runtime/artifacts/bin/Syste"..., "--depsfile", "/mnt/runtime/artifacts/bin/Syste"..., "/mnt/runtime/artifacts/bin/Syste"..., "System.Net.Requests.Tests, Versi"..., "System.Net.Tests.AuthenticationM"..., "<CustomTargetNameDictionary_Vali"..., "/tmp/3ibo3kk5.gfq"], 0x7e90051de0 /* 13 vars */) = 0
16637 --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=16664, si_uid=1001, si_status=42, si_utime=10, si_stime=6} ---
16630 waitid(P_ALL, 0, {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=16664, si_uid=1001, si_status=42, si_utime=0, si_stime=0}, WNOHANG|WEXITED|WNOWAIT, NULL) = 0
16630 wait4(16664, [{WIFEXITED(s) && WEXITSTATUS(s) == 42}], WNOHANG, NULL) = 16664
16630 waitid(P_ALL, 0, {}, WNOHANG|WEXITED|WNOWAIT, NULL) = 0
16678 execve("/mnt/runtime/artifacts/bin/testhost/net7.0-Linux-Debug-arm64/dotnet", ["/mnt/runtime/artifacts/bin/testh"..., "exec", "--runtimeconfig", "/mnt/runtime/artifacts/bin/Syste"..., "--depsfile", "/mnt/runtime/artifacts/bin/Syste"..., "/mnt/runtime/artifacts/bin/Syste"..., "System.Net.Requests.Tests, Versi"..., "System.Net.Tests.AuthenticationM"..., "<Register_Unregister_ModuleCount"..., "/tmp/tuctrifw.tyv"], 0x7e9002e400 /* 13 vars */) = 0
16689 execve("/mnt/runtime/artifacts/bin/testhost/net7.0-Linux-Debug-arm64/dotnet", ["/mnt/runtime/artifacts/bin/testh"..., "exec", "--runtimeconfig", "/mnt/runtime/artifacts/bin/Syste"..., "--depsfile", "/mnt/runtime/artifacts/bin/Syste"..., "/mnt/runtime/artifacts/bin/Syste"..., "System.Net.Requests.Tests, Versi"..., "System.Net.Tests.HttpWebRequestT"..., "<DefaultMaximumResponseHeadersLe"..., "/tmp/5edbvy3h.1x1"], 0x7e8c07b110 /* 13 vars */) = 0
16637 --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=16678, si_uid=1001, si_status=42, si_utime=12, si_stime=9} ---
16690 execve("/mnt/runtime/artifacts/bin/testhost/net7.0-Linux-Debug-arm64/dotnet", ["/mnt/runtime/artifacts/bin/testh"..., "exec", "--runtimeconfig", "/mnt/runtime/artifacts/bin/Syste"..., "--depsfile", "/mnt/runtime/artifacts/bin/Syste"..., "/mnt/runtime/artifacts/bin/Syste"..., "System.Net.Requests.Tests, Versi"..., "System.Net.Tests.HttpWebRequestT"..., "<GetResponseAsync_ParametersAreC"..., "/tmp/ad05mcf0.phu", "False"], 0x7e8403e720 /* 13 vars */) = 0
16630 waitid(P_ALL, 0, {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=16678, si_uid=1001, si_status=42, si_utime=0, si_stime=0}, WNOHANG|WEXITED|WNOWAIT, NULL) = 0
16630 wait4(16678, [{WIFEXITED(s) && WEXITSTATUS(s) == 42}], WNOHANG, NULL) = 16678
16630 waitid(P_ALL, 0, {}, WNOHANG|WEXITED|WNOWAIT, NULL) = 0
16691 execve("/mnt/runtime/artifacts/bin/testhost/net7.0-Linux-Debug-arm64/dotnet", ["/mnt/runtime/artifacts/bin/testh"..., "exec", "--runtimeconfig", "/mnt/runtime/artifacts/bin/Syste"..., "--depsfile", "/mnt/runtime/artifacts/bin/Syste"..., "/mnt/runtime/artifacts/bin/Syste"..., "System.Net.Requests.Tests, Versi"..., "System.Net.Tests.AuthenticationM"..., "<CredentialPolicy_Roundtrip>b__6"..., "/tmp/jqxqxkup.irr"], 0x7e9002d320 /* 13 vars */) = 0
16638 --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=16689, si_uid=1001, si_status=42, si_utime=11, si_stime=6} ---
16630 waitid(P_ALL, 0, {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=16689, si_uid=1001, si_status=42, si_utime=0, si_stime=0}, WNOHANG|WEXITED|WNOWAIT, NULL) = 0
16630 wait4(16689, [{WIFEXITED(s) && WEXITSTATUS(s) == 42}], WNOHANG, NULL) = 16689
16630 waitid(P_ALL, 0, {}, WNOHANG|WEXITED|WNOWAIT, NULL) = 0
16637 --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=16691, si_uid=1001, si_status=42, si_utime=11, si_stime=11} ---
16630 waitid(P_ALL, 0, {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=16691, si_uid=1001, si_status=42, si_utime=0, si_stime=0}, WNOHANG|WEXITED|WNOWAIT, NULL) = 0
16630 wait4(16691, [{WIFEXITED(s) && WEXITSTATUS(s) == 42}], WNOHANG, NULL) = 16691
16630 waitid(P_ALL, 0, {}, WNOHANG|WEXITED|WNOWAIT, NULL) = 0
16711 execve("/mnt/runtime/artifacts/bin/testhost/net7.0-Linux-Debug-arm64/dotnet", ["/mnt/runtime/artifacts/bin/testh"..., "exec", "--runtimeconfig", "/mnt/runtime/artifacts/bin/Syste"..., "--depsfile", "/mnt/runtime/artifacts/bin/Syste"..., "/mnt/runtime/artifacts/bin/Syste"..., "System.Net.Requests.Tests, Versi"..., "System.Net.Tests.AuthenticationM"..., "<Register_UnregisterByScheme_Mod"..., "/tmp/hzkvicop.2dw"], 0x7e9002d3c0 /* 13 vars */) = 0
16728 --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=16711, si_uid=1001, si_status=42, si_utime=13, si_stime=5} ---
16630 waitid(P_ALL, 0, {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=16711, si_uid=1001, si_status=42, si_utime=0, si_stime=0}, WNOHANG|WEXITED|WNOWAIT, NULL) = 0
16630 wait4(16711, [{WIFEXITED(s) && WEXITSTATUS(s) == 42}], WNOHANG, NULL) = 16711
16630 waitid(P_ALL, 0, {}, WNOHANG|WEXITED|WNOWAIT, NULL) = 0
16718 --- SIGRTMIN {si_signo=SIGRTMIN, si_code=SI_TKILL, si_pid=16663, si_uid=1001} ---
16719 --- SIGRTMIN {si_signo=SIGRTMIN, si_code=SI_TKILL, si_pid=16663, si_uid=1001} ---
16728 --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=16663, si_uid=1001, si_status=42, si_utime=95, si_stime=32} ---
16630 waitid(P_ALL, 0, {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=16663, si_uid=1001, si_status=42, si_utime=0, si_stime=0}, WNOHANG|WEXITED|WNOWAIT, NULL) = 0
16630 wait4(16663, [{WIFEXITED(s) && WEXITSTATUS(s) == 42}], WNOHANG, NULL) = 16663
16630 waitid(P_ALL, 0, {}, WNOHANG|WEXITED|WNOWAIT, NULL) = 0
....
```


I'm running the test suite in loop but as usually no repro.  It is interesting that the dump above has `_usesTerminal=1`.  I was thinking about differences in CI and one is that the test suite is executed from background Helix daemon vs running in foreground from interactive shell. I don't think it _should_ make much difference for process management but the process code has some specific processing.

**Comment by tmds** [comment]
The process exit handling code hasn't changed much in the past years. I don't expect it to have a bug as it has been used probably millions of times in various situations without users reporting the `errno = 10` case.

I think we should treat this as a glitch unless it reproduces more.

**Comment by tmds** [comment]
> However it fails somewhat often in CI.

I thought this was a single occurrence. Since when did it start failing in CI?

**Comment by wfurt** [comment]
It feels like it start happening more in last few weeks. But as far as I can tell the https://github.com/dotnet/runtime/issues/55618 was never solved, right? 
There was #73972 so it is hard to know all the details without going through the logs by hand. (Kusto reports same result)
With the corruption we would generally get crash but the `waitpid` issue is different and consistent. 

I see ~ 200+ failures in last two weeks. Mono was disabled in  #74668 (see https://github.com/dotnet/runtime/issues/74667)

**Comment by tmds** [comment]
https://github.com/dotnet/runtime/issues/55618 was reported over a year ago, and it was closed after the issue had no longer occurred for 3 weeks. Since then it wasn't reported until this issue was created this week.

Having an idea of when it started to occur more can give us a clue as to what change triggered it.

> I see ~ 200+ failures in last two weeks. Mono was disabled in https://github.com/dotnet/runtime/pull/74668 (see https://github.com/dotnet/runtime/issues/74667)

We've only seen this on arm64?

Does it occur less with coreclr than with mono?

The code asks the kernel if there are children that terminated using `waitid`:
```
16630 waitid(P_ALL, 0, {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=16711, si_uid=1001, si_status=42, si_utime=0, si_stime=0}, WNOHANG|WEXITED|WNOWAIT, NULL) = 0
```
If there is one, it passes the `pid` (from `si_pid`) to reap the child:
```
16630 wait4(16711, [{WIFEXITED(s) && WEXITSTATUS(s) == 42}], WNOHANG, NULL) = 16711
```

In between these calls the child suddenly goes missing.

So either something has messed up the `pid`, or something else has reaped the child in the meanwhile.

If it's something else reaping the child, it's expected to be a native library making a blocking `waitpid(-1, &status, 0)` call (or `wait(&status)`) on a thread. This thread will race with .NET signal handling as soon as the child terminates.

If there is no native library doing this, then somehow the `pid` changed.
Or there is a bug in the kernel.

Note that the signal handling is running on is native pthread. It's not a managed thread.

**Comment by wfurt** [comment]
I see it recently only on ARM. Some of the older cases are gone e.g. there is note of crash on Kusto but we no longer have console so we cannot check if the pattern is the same. I searched our code base (and msquic) and I did not see any direct uses of `wait*`. So I'm running out of ideas. 

Now, since you mentioned the `-1`: #72830 is another mysterious case when Socket's `SafeHandle` turns into -1. There is also no local repro so far and insight is minimal.

**Comment by wfurt** [comment]
actually, the runs are containers. I'm wondering if this is applicable:
https://github.com/dotnet/runtime/blob/3afb168cbfc7edefc66a4d6c052c777d94b58ff0/src/native/libs/System.Native/pal_signal.c#L349-L354

That would eventually call 
```c#
 pid = Interop.Sys.WaitPidExitedNoHang(-1, out _);
```

**Comment by jkotas** [comment]
> Now, since you mentioned the -1: https://github.com/dotnet/runtime/issues/72830 is another mysterious case when Socket's SafeHandle turns into -1.

Yes, both of these can have the case root cause.

Notice that both of these issues have been only seen in  System.Net.Requests.Tests.

One possible explanation is that a method passes a pointer to a stack location to some asynchronous logic. The method goes out of scope, and some completely unrelated code starts running. And finally, the asynchronous logic wakes up and overwrites the location on the stack that is used for something completely different by now.

Does it ring any bells? Do we have a code in sockets that is passing a pointer to a stack location to some asynchronous logic?

**Comment by wfurt** [comment]
I will take a look. https://github.com/dotnet/runtime/issues/72830 happens in System.Net.Mail. That has really only one functional change recently and I don't see how that would cause either of the problems.

**Comment by jkotas** [comment]
> I will take a look. https://github.com/dotnet/runtime/issues/72830 happens in System.Net.Mail.

Right, but the mysterious crash is still in sockets.

**Comment by tmds** [comment]
> actually, the runs are containers. I'm wondering if this is applicable:

It isn't. It is applicable only for the main process in the container (pid 1). You can see `reapAll = false` in your debugger output.

I don't see any reports of this on x64.

Was Mono disabled because it happens more on Mono than it does on coreclr?

**Comment by steveisok** [comment]
> Was Mono disabled because it happens more on Mono than it does on coreclr?

On mono Linux arm64, it was happening on every ci run.

**Comment by akoeplinger** [comment]
I figured out an issue with one test that fixes the crash on Mono for me, there's a good chance that it helps this one too: https://github.com/dotnet/runtime/pull/74994

**Comment by akoeplinger** [comment]
Turns out this is still happening on Mono even after https://github.com/dotnet/runtime/pull/74994 so this was an unrelated issue.

Interestingly I am able to repro it locally in the Docker container we also use on Helix fairly reliably when it is the first thing that runs in the new container instance, e.g. I start the container and then run the repro -> it crashes. If however it works the first time then I can run the repro in a loop for 15mins without any crash 😄

**Comment by akoeplinger** [comment]
I added logging the PID to the "Error while reaping child" error message and also log each RemoteExecutor-PID and the crash doesn't seem to be specific to one single test.

**Comment by tmds** [comment]
You are reproducing this on arm64 using Mono?

Can we make some correlation between the failures and the container images being used?

The last change (https://github.com/dotnet/runtime/pull/73963) was for enabling Quic support. That sounds like it could affect `System.Net.Requests.Tests`.

**Comment by akoeplinger** [comment]
@tmds yes this is using Mono. Some of the threads are indeed inside msquic code, see the full thread dump I get due to the FailFast here: https://gist.github.com/akoeplinger/b5768ec63cc36a4c672085c621690853.

**Comment by akoeplinger** [comment]
I tried using the new Docker container and removing `/usr/lib/aarch64-linux-gnu/libmsquic.so.2` as the first step.
It ran for 10 minutes without a crash.

Adding msquic back and it crashed after a couple iterations. So I think Quic is the most likely culprit.

**Comment by tmds** [comment]
@akoeplinger can you share how you build the repo and run the test to reproduce the issue?
Will you or @wfurt investigate further?

**Comment by akoeplinger** [comment]
Here's what I did:

1. Clone the repo and run `./build.sh mono+libs -rc Release -c Debug` (that's the config on PR builds, no idea if it's relevant)
2. Run this to build the tests once on the host: `./dotnet.sh build /t:Test /p:RuntimeFlavor=mono /p:MonoConfiguration=Release src/libraries/System.Net.Requests/tests/` 
3. Run `docker run -v $(pwd):/repo -i mcr.microsoft.com/dotnet-buildtools/prereqs:debian-11-helix-arm64v8-20220820185253-06f234f bash < repro.sh`

The contents of repro.sh are:
```
pushd /repo/artifacts/bin/System.Net.Requests.Tests/Debug/net7.0
/repo/artifacts/bin/testhost/net7.0-Linux-Debug-arm64/dotnet exec --runtimeconfig System.Net.Requests.Tests.runtimeconfig.json --depsfile System.Net.Requests.Tests.deps.json xunit.console.dll System.Net.Requests.Tests.dll -xml testResults.xml -nologo -notrait category=OuterLoop -notrait category=failing
exit $?
```

This will run the tests and then exit. Note that for some reason it seems there's an error writing testResult.xml due to permissions so the exit code will be 4. I just used that and run the Docker command in a while loop.

Environment:
Ubuntu 18.04 arm64 with 6 cores, running on an M1 MacBook via Parallels.

I'm happy to add more logging or run experiments etc. but I don't think I can meaningfully investigate the issue since I don't know Quic/System.Net.Requests

**Comment by danmoseley** [comment]
> The corrupted state here is akin to calling free on a pointer that you do not own.

should this be set to milestone 7.0?

**Comment by wfurt** [comment]
I will try it with the fresh docker. I was talking to @MattGal last week how Helix orchestrates the test runs in docker but I was out on Friday and yesterday.

**Comment by wfurt** [comment]
I could reproduce it with `docker run`. I did not see any signs that either `MsQuic` or `Process` is misbehaving. (may be just timing difference) Interestingly, I could not reproduce it with `dotnet exec` (e.g. variant of running container with more tools)
I'll continue with investigation and I may try it with coreclr to see if `Mono` really matters.

**Comment by karelz** [comment]
This is heavily impacting CI -- most failures are on arm64 Mono -- can someone familiar with Mono please take a look ASAP? @akoeplinger @steveisok

**Comment by akoeplinger** [comment]
Mono Linux arm64 tests were disabled with #75266

**Comment by danmoseley** [comment]
As we have reason to suspect a product bug, should this be milestone 7.0?

**Comment by jkotas** [comment]
> As we have reason to suspect a product bug, should this be milestone 7.0?

Yes, I think so.

**Comment by wfurt** [comment]
This is interesting. I can reproduce with `debian-11-helix-arm64v8-20220820185253-06f234f ` @akoeplinger  used. But not with `debian-11-helix-arm64v8-20220908131932-59ee6b3` that has MsQuic 2.1.1. (note that begin of the numbers is date when the image was built)

Is there any known issue fixed in msquic in ~last month @nibanks that could cause memory corruption? 
I can see that quince and noquic can cause timing differences but it is curious that the problem disappears with updated image.

**Comment by wfurt** [comment]
I have finally some smoking gun. 
@tmds was right, something is easing out pids.

```
31    02:15:15 +++ exited with 42 +++
25    02:15:15 --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=31, si_uid=1000, si_status=42, si_utime=51, si_stime=7} ---
14    02:15:15 waitid(P_ALL, 0, {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=31, si_uid=1000, si_status=42, si_utime=0, si_stime=0}, WNOHANG|WEXITED|WNOWAIT, NULL) = 0
21    02:15:15 clone(child_stack=NULL, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLD, child_tidptr=0x7f988ee280) = 45
21    02:15:15 wait4(-1, [{WIFEXITED(s) && WEXITSTATUS(s) == 42}], 0, NULL) = 31
21    02:15:15 clone(child_stack=NULL, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLD, child_tidptr=0x7f988ee280) = 46
21    02:15:15 wait4(-1,  <unfinished ...>
46    02:15:15 exit_group(0)            = ?
45    02:15:15 exit_group(0)            = ?
46    02:15:15 +++ exited with 0 +++
21    02:15:15 <... wait4 resumed>[{WIFEXITED(s) && WEXITSTATUS(s) == 0}], 0, NULL) = 46
21    02:15:15 --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=46, si_uid=1000, si_status=0, si_utime=0, si_stime=2} ---
21    02:15:15 clone(child_stack=0x7f6e4a3ab0, flags=CLONE_VM|CLONE_FS|CLONE_FILES|CLONE_SIGHAND|CLONE_THREAD|CLONE_SYSVSEM|CLONE_SETTLS|CLONE_PARENT_SETTID|CLONE_CHILD_CLEARTID, parent_tid=[47], tls=0x7f6e4a48b0, child_tidptr=0x7f6e4a4280) = 47
45    02:15:15 +++ exited with 0 +++
21    02:15:15 --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=45, si_uid=1000, si_status=0, si_utime=0, si_stime=4} ---
14    02:15:15 wait4(31, 0x7f9a0b7504, WNOHANG, NULL) = -1 ECHILD (No child processes)
```
14 is the signal handler and 20 is another .NET thread. I fiddle with that more and I got `strace` that supports stack traces.

```
29    05:58:40 wait4(-1, [{WIFEXITED(s) && WEXITSTATUS(s) == 42}], 0, NULL) = 32
 > /lib/aarch64-linux-gnu/libc-2.31.so(wait4+0x80) [0xa7b80]
 > /repo/libs/liblttng-ust.so.0(get_map_shm+0x454) [0xff84]
 > /repo/libs/liblttng-ust.so.0(lttng_ust_init.part.5+0x5e8) [0x10b48]
 > /lib/aarch64-linux-gnu/ld-2.31.so(_dl_rtld_di_serinfo+0x593c) [0xe93c]
 > /lib/aarch64-linux-gnu/ld-2.31.so(_dl_rtld_di_serinfo+0x5a44) [0xea44]
 > /lib/aarch64-linux-gnu/libc-2.31.so(_dl_catch_exception+0xd4) [0x10fb74]
 > /lib/aarch64-linux-gnu/ld-2.31.so(_dl_find_dso_for_object+0xb10) [0x12c30]
 > /lib/aarch64-linux-gnu/libc-2.31.so(_dl_catch_exception+0x74) [0x10fb14]
 > /lib/aarch64-linux-gnu/ld-2.31.so(_dl_find_dso_for_object+0x180) [0x122a0]
 > /lib/aarch64-linux-gnu/libdl-2.31.so() [0x1264]
 > /lib/aarch64-linux-gnu/libc-2.31.so(_dl_catch_exception+0x74) [0x10fb14]
 > /lib/aarch64-linux-gnu/libc-2.31.so(_dl_catch_error+0x40) [0x10fbe0]
 > /lib/aarch64-linux-gnu/libdl-2.31.so(dlerror+0x310) [0x1b10]
 > /lib/aarch64-linux-gnu/libdl-2.31.so(dlopen+0x54) [0x1304]
 > /usr/lib/aarch64-linux-gnu/libmsquic.so.2(CxPlatSystemLoad+0xe4) [0x97074]
 > /usr/lib/aarch64-linux-gnu/libmsquic.so.2(MsQuicLibraryLoad+0x50) [0x37110]
```

and sure enough lttng is using `wait()` instead of `waitpid()`. This will happily collect children from another threads within same process.
https://github.com/lttng/lttng-ust/blob/8e95eaadb72a1a97486f15cc128ff6ed2fcd0196/src/lib/lttng-ust/lttng-ust-comm.c#L1583-L1597

I built `lttng` from source and with some more instrumentation I verified that our missing pid is eaten by them:
```
SystemNative_WaitPidExitedNoHang:719: SystemNative_WaitPidExitedNoHang returned -1 for 32 status 10 No child processes
TryReapChild: Failed to reap 32 from 11
get_wait_shm:1237: wait finished with 32 and 10752
```

while I searched msquic and did not find any suspicious code it is invoked indirectly through library msquic is using. 
We can submit fix to lttng bit it will take a while and we cannot depend on it. 

The handling in `Process` seems overstrict. As far as I can tell there is no corruption and crashing the process seems excessive. Throwing (so caller can catch and deal with the error) or returning sentinel result may be better. 

It also feels like we should do do tracing unless asked for.  I used `QUIC_LTTng=0` @jkotas  suggested and I verified that issue disappears. Perhaps `lttng` should be called only if `QUIC_LTTng=1` or we should add API to control it?
Any thoughts on this @nibanks ?

We could possibly split libraries into two packages: 
`libmsquic` only  containing `libmsquic.so.2` and move `libmsquic.lttng.so.2.1.1` to separate package? 

We could also re-shuffle the tests so it loads `Quic` early before we start tests using RemoteExecutor. 
But this feels fragile and solves only test runs, not the product issue.

**Comment by tmds** [comment]
> I have finally some smoking gun.

Great!

> The handling in Process seems overstrict. As far as I can tell there is no corruption and crashing the process seems excessive. Throwing (so caller can catch and deal with the error) or returning sentinel result may be better.

If we can't reap the process, we can't determine it's exit code.
Now we `FailFast` when we detect unexpected state.

If we'd fake an exit code, we're corrupting the state.
If we'd only throw if someone gets the exit code, we're still introducing an exception where none is expected. It may become less likely the exception crashes the process, but the issue still exist, and it can still crash the process.

The `FailFast` makes us look deeper, and find and fix the root cause.

@wfurt are you logging an issue with lttng? and maybe send a fix?

**Comment by jkotas** [comment]
> Perhaps lttng should be called only if QUIC_LTTng=1 or we should add API to control it?

LTTng tracing is meant to be initialized by default so that one can start the tracing session remotely without changing anything else. It matches what we do for coreclr itself in https://github.com/dotnet/runtime/blob/main/src/coreclr/pal/src/misc/tracepointprovider.cpp#L63 .

**Comment by jkotas** [comment]
Our tests have a specific pattern that is prone to hit the lttng bug: The tests are initializing Quic on one thread while there is a remote executor exec storm of short-lived processes on other threads. I think that it is unlikely for real world applications out there to have the same pattern. 

Until we get lttng fix in place, we may want to set `QUIC_LTTng=0` in our CI docker images.

**Comment by wfurt** [comment]
This is not just the docker, right? Since this is showing up so far only on ARM64, we may choose to update only that. 
I put up #75424 with alternative approach to load msquic early before tests runs. That also seems sufficient.

I'm working on issue for LTTNG. They don't use GitHub for tracking and my registration to submit issue is pending for approval. I look at the code again and changing `wait` to `waitpid` should be safe and it it should fix the observed behavior.

**Comment by karelz** [comment]
@wfurt do we understand why it is happening only on System.Net.Request on not on other test suites?
Will we need to initialize msquic in more test suites in future?

**Comment by karelz** [comment]
Also, does it impact also release/7.0 branch? Should we backport it as test-only change?

**Comment by jkotas** [comment]
> @wfurt do we understand why it is happening only on System.Net.Request on not on other test suites?
> Will we need to initialize msquic in more test suites in future?

Any test that launches child processes and ends up loading Quic is prone to hitting this issue. I expect that we will see this problem in a few more tests over time. 

An alternative big hammer fix that would address it everywhere would be to set `QUIC_LTTng=0` in the environment. I think we try the proposed fix and see how it works, and only go back to the big hammer fix if we get signal that the problem is wide-spread.

>  does it impact also release/7.0 branch? Should we backport it as test-only change?

Yes.

**Comment by wfurt** [comment]
Note this is initialization race. So when Quic is initialized befe test runs and child processes it is OK. That for example happens when we conditional tests where XUnit evaluates conditions before test run.

**Comment by wfurt** [comment]
reopening for 7.0

**Comment by karelz** [comment]
@wfurt can you please also re-enable the disabled test from PR #75266

**Comment by wfurt** [comment]
That was part of #75424 unless I missed something. @jkotas already asked for it during review.

**Comment by wfurt** [comment]
https://github.com/dotnet/runtime/pull/75266 does not seems to make it to 7.0 so that part is missing the back-port PR.

**Comment by karelz** [comment]
Yep, sorry - I looked accidentally only at 7.0 change where it was not disabled ... I updated to post and removed the disabled-test label.

**Comment by karelz** [comment]
Fixed in main (8.0) in PR #75424 and in 7.0 (RC2) in PR #75466.

**Comment by tmds** [comment]
For reference, the lttng issue created by @wfurt: https://bugs.lttng.org/issues/1359.
